multigeiger/speaker.cpp

```diff
@@ -65,11 +65,11 @@ void IRAM_ATTR isr_audio() {
     led = *p++;
     duration_ms = *p++;
     isr_sequence = p;
+    portEXIT_CRITICAL_ISR(&mux_audio);
+  } else {
+    portEXIT_CRITICAL_ISR(&mux_audio);
+    return;
   }
-  portEXIT_CRITICAL_ISR(&mux_audio);
-
-  if (!isr_sequence)
-    return;  // nothing to do
```

Comment:
2) If now we have the last part of the sequence, isr_sequence is NULL, however the parameters (frequency, volume, led, duration) have not yet been interpreted, i.e. the sequences have not been NULLed yet. But nevertheless we return here. So p / isr_sequence are not the right criteria to return.

Reply:
i was suggesting to use !p here (p being assigned in the mutex block right above).
p is not NULL if there is something to do still.

Reply:
Yes, but imo !p has the same issue, being NULL after the last sequence part has been parsed, despite the corresponding paramaters have not yet been processed.